stl/inc/ranges

```diff
@@ -620,6 +620,132 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
+    // CLASS TEMPLATE ranges::_Non_propagating_cache
+    template <class _Ty>
+    concept _Objectable = is_object_v<_Ty>;
+
+    // clang-format off
+    template <_Objectable _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20_DYNALLOC ~_Non_propagating_cache() {

```

Comment:
We should have a separate `~_Non_propagating_cache() requires is_trivially_destructible_v<_Ty> = default;`.

Reply:
There is a different class specialization for trivially destructible types.

Reply:
> There is a different class specialization for trivially destructible types.

There's a different specialization for trivially-destructible default-constructible types; this specialization will be used for trivially-destructible non-default-constructible types.

Reply:
Why make this specialization also require default constructability? We should be able to remove `_Engaged` member for all trivially destructible types, given that it only controls invocation of `_Val`'s destructor (which does nothing if it's trivial).

Reply:
I fear we actually need both specializations